File: C-Plus-Plus/cp/waveArray.cpp
/*	Given an array of integers, return the array which is sorted into a wave like array.
 *	In other words, arrange the elements into a sequence such that a1 >= a2 <= a3 >= a4 <= a5...
 *
 *	Example:
 *	Given [1, 2, 3, 4]
 *	One possible answer : [2, 1, 4, 3]
 *	Another possible answer : [4, 1, 3, 2]
 */

#include <bits/stdc++.h>
using namespace std;

vector<int> waveArray(int n, vector<int> a){
    sort(a.begin(), a.end());
    for(int curr = 0; curr < n - 1; curr += 2){
        swap(a[curr], a[curr + 1]);
    }
    return a;
}

int main(){
    int size, val;
    vector <int> v;
    cout << "Enter the size of the array: ";
    cin >> size;
			    
    cout << "Enter the elements of the array: " << endl;
    for(int i = 0; i < size; i++){
        cin >> val;
        v.push_back(val);
    }
				        
    v = waveArray(size, v);
    cout << "Wave Array: ";			        
    for(int i = 0; i < size; i++)
        cout << v[i] << " ";
    cout << endl;
}

/*
 * 	Sample Output:
 *
 *  	Enter the size of the array: 5
 *      Enter the elements of the array:
 *      1
 *      2
 *      3
 *      4
 *      5
 *      Output:
 *      Wave Array: 2 1 4 3 5
 *
 *      Time Complexity: O(nlogn)
 *      Space Complexity: O(1)
 *
 */

File: Python/other/Wave_Array.py
"""
Sorting the array into a wave-like array.
For eg: arr[] = {1,2,3,4,5}
Output: 2 1 4 3 5
"""


def convertToWave(length,array):
    for i in range(0, length - length%2, 2):               #swapping alternatively
        temp=array[i]
        array[i]=array[i+1]
        array[i+1]=temp
    return array

arr = list(map(int,input("Enter the elements of the array : ").split()))
N = len(arr)    # length of the array

arr = convertToWave(N,arr)

print("Wave Array: ")
for el in arr:
  print(el, end=" ")
   
    
"""
Time complexity : O(1) 
Space complexity : O(1) 

INPUT:-
Enter the elements of the array : 1 2 3 4 5 6 7
OUTPUT:-
Wave Array: 2 1 4 3 6 5 7 

"""
